cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -188,6 +201,81 @@ public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile
+
+    ClassLoader pluginClassLoader = pluginInstantiator.getArtifactClassLoader(artifact.toArtifactId());
+
+    Class pluginClassLoaded = pluginClassLoader.loadClass(pluginClass.getClassName());
+
+    Method[] methods = pluginClassLoaded.getDeclaredMethods();
+    for (Method method : methods) {
+      Path pathAnnotation = method.getAnnotation(Path.class);
+      // method should have path annotation else continue
+      if (pathAnnotation == null || !pathAnnotation.value().equals(methodName)) {
+        continue;
+      }
+
+      // make configField name as null, as we might want to call these methods with incomplete config.
+      PluginClass pluginClassWithoutConfig = new PluginClass(pluginClass.getType(), pluginClass.getName(),
+                                                             pluginClass.getDescription(), pluginClass.getClassName(),
+                                                             null, pluginClass.getProperties(),
+                                                             pluginClass.getEndpoints());
+      co.cask.cdap.api.plugin.Plugin plugin =
+        new co.cask.cdap.api.plugin.Plugin(artifact.toArtifactId(), pluginClassWithoutConfig,
+                                           PluginProperties.builder().build());
+      try {
+        Object result;
+        if (method.getParameterTypes().length == 2 &&
+          method.getParameterTypes()[1].equals(EndpointPluginContext.class)) {
+          result = method.invoke(pluginInstantiator.newInstance(plugin),
+                                 GSON.fromJson(methodParameter, method.getParameterTypes()[0]),
+                                 endpointPluginContext);
+        } else {
+          result = method.invoke(pluginInstantiator.newInstance(plugin),
+                                 GSON.fromJson(methodParameter, method.getParameterTypes()[0]));
+        }
+
+        // serialize the result and return
+        return GSON.toJson(result, method.getReturnType());
+      } catch (InvocationTargetException e) {
+        LOG.error("Got Exception when invoking plugin method {}", e);
+        if (e.getCause() instanceof NotFoundException) {
+          throw new NotFoundException(e.getCause());
+        } else if (e.getCause() instanceof BadRequestException) {
+          throw new BadRequestException(e.getCause());
+        } else {
+          throw Throwables.propagate(e);
+        }
+      } catch (Throwable t) {
+        throw new InvalidArtifactException(String.format(
+          "Class could not be found while inspecting artifact for plugins. " +
+            "Please check dependencies are available, and that the correct parent artifact was specified. " +
+            "Error class: %s, message: %s.", t.getClass(), t.getMessage()), t);
+      }
+    }
+
+    // cannot find the plugin or plugin method to invoke.
+    throw new NotFoundException("Could not find the plugin with matching name and type to invoke");
```

Comment:
this is not an accurate error message.  You couldn't find the method to invoke.  But shouldn't we know this immediately because we store it in PluginClass?